File: clients/python/debugging/camera.py
from model.vec2 import Vec2
from stream_wrapper import StreamWrapper

class Camera:
    """Camera state"""

    __slots__ = ("center","rotation","attack","fov",)

    center: Vec2
    rotation: float
    attack: float
    fov: float

    def __init__(self, center: Vec2, rotation: float, attack: float, fov: float):
        self.center = center
        """Center"""
        self.rotation = rotation
        """Rotation"""
        self.attack = attack
        """Attack angle"""
        self.fov = fov
        """Vertical field of view"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Camera":
        """Read Camera from input stream
        """
        center = Vec2.read_from(stream)
        rotation = stream.read_double()
        attack = stream.read_double()
        fov = stream.read_double()
        return Camera(center, rotation, attack, fov)
    
    def write_to(self, stream: StreamWrapper):
        """Write Camera to output stream
        """
        self.center.write_to(stream)
        stream.write_double(self.rotation)
        stream.write_double(self.attack)
        stream.write_double(self.fov)
    
    def __repr__(self):
        return "Camera(" + \
            repr(self.center) + \
            ", " + \
            repr(self.rotation) + \
            ", " + \
            repr(self.attack) + \
            ", " + \
            repr(self.fov) + \
            ")"

File: clients/javascript/debugging/camera.js
const Vec2 = require.main.require('./model/vec2');
/**
 * Camera state
 */
class Camera {
    /**
     * Center
     */
    center;
    /**
     * Rotation
     */
    rotation;
    /**
     * Attack angle
     */
    attack;
    /**
     * Vertical field of view
     */
    fov;

    constructor(center, rotation, attack, fov) {
        this.center = center;
        this.rotation = rotation;
        this.attack = attack;
        this.fov = fov;
    }

    /**
     * Read Camera from input stream
     */
    static async readFrom(stream) {
        let center;
        center = await Vec2.readFrom(stream);
        let rotation;
        rotation = await stream.readDouble();
        let attack;
        attack = await stream.readDouble();
        let fov;
        fov = await stream.readDouble();
        return new Camera(center, rotation, attack, fov);
    }

    /**
     * Write Camera to output stream
     */
    async writeTo(stream) {
        let center = this.center;
        await center.writeTo(stream);
        let rotation = this.rotation;
        await stream.writeDouble(rotation);
        let attack = this.attack;
        await stream.writeDouble(attack);
        let fov = this.fov;
        await stream.writeDouble(fov);
    }
}
module.exports = Camera